### PR TITLE
Added support for multiple --from-env flags

### LINKED ISSUE #1033: kubectl create configmap: Support multiple --from-env-file flags
**What would you like to be added**:
I would like an error to be raised if multiple `--from-env-file` flags are specified in the `kubectl create configmap` command.

**Why is this needed**:
If multiple `--from-env-file` flags are specified, the last one overrides the previous ones. Therefore, --from-env-file is useless and ambiguous.

[Source](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/)
> Caution: When passing --from-env-file multiple times to create a ConfigMap from multiple data sources, only the last env-file is used.

### PR DESCRIPTION
**Pull request by lauchokyip**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Added support for multiple --from-env flags

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubectl/issues/1033

#### Does this PR introduce a user-facing change?

```release-note
Adding support for multiple --from-env-file flags
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go [hunk 1]
```diff
@@ -414,8 +458,11 @@ func TestCreateConfigMap(t *testing.T) {
 					defer teardown()
 				}
 			}
+			err := configMapOptions.Validate()
```
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go:461]: Why is this added?

**Review by KnVerey — COMMENTED** [review]
Great work! I have a few small questions/quibbles, but I tried it out locally and it is working nicely.

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go (hunk 1 above)
Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go:461]: I was the one that remove the generator of the ConfigMap (https://github.com/pulls?q=is%3Apr+author%3Alauchokyip+archived%3Afalse+is%3Aclosed) and when I wrote the test cases for example this test cases https://github.com/kubernetes/kubernetes/blob/a45aeb626c7f2303c49466ae52833cd410cf88f2/staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go#L386 `Validate()` would catch the error here https://github.com/kubernetes/kubernetes/blob/a45aeb626c7f2303c49466ae52833cd410cf88f2/staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap.go#L208, same as what I did for `create_secret_test.go` https://github.com/kubernetes/kubernetes/blob/a45aeb626c7f2303c49466ae52833cd410cf88f2/staging/src/k8s.io/kubectl/pkg/cmd/create/create_secret_test.go#L480

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go [hunk 2]
```diff
@@ -429,20 +476,28 @@ func TestCreateConfigMap(t *testing.T) {
 	}
 }
 
-func setupEnvFile(lines ...string) func(*testing.T, *ConfigMapOptions) func() {
+func setupEnvFile(filenames []string, lines [][]string) func(*testing.T, *ConfigMapOptions) func() {
```
[the remark this answers is not in the document]
Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go:479]: I believe it's needed because after setting up the filenames, we would want to the mapping from `file.env` to `cmenv` over [here](https://github.com/kubernetes/kubernetes/blob/a45aeb626c7f2303c49466ae52833cd410cf88f2/staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go#L443)

[the remark this answers is not in the document]
Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go]: no,  it's no needed 😕

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go (hunk 1 above)
Comment (KnVerey) [reply staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go:461]: What I still don't understand is why this PR makes this change necessary. If the errors the cases you point to are meant to hit are in `Validate` and only `createConfigMap` is being called on master today, then how are the tests passing on master? Are they hitting errors other than the ones they're meant to? Incidentally, that possibility is one reason why I personally prefer to make error expectations strings to match on rather than booleans, to provide better assurances that the error that happens is the one I expect.

To be clear, I'm not opposed to this change--just want to make sure I understand it.

Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go:461]: Will take a look into it

Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go:461]: Turn out the error happened during `createConfigMap` 
For instance the test case `create_invalid_configmap_too_many_args`, instead of throwing the error `from-env-file cannot be combined with from-file or from-literal` it actually tries to create a config map from the file and it cannot read the file therefore throws an error

I agree comparing string error is way better so I will make a change from that

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go [hunk 3]
```diff
@@ -352,104 +396,130 @@ func TestCreateConfigMap(t *testing.T) {
...
 			}
+			err := configMapOptions.Validate()
+			if err == nil {
+				configMap, err = configMapOptions.createConfigMap()
+			}
 
-			configMap, err := configMapOptions.createConfigMap()
 			if !test.expectErr && err != nil {
-				t.Errorf("test %s, unexpected error: %v", name, err)
+				t.Errorf("unexpected error: %v", err)
```
Comment (lauchokyip) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go:477]: removed the `name` because if the unit test fails it would already output the name so in my opinion it's repetitive and confusing

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap.go [hunk 4]
```diff
@@ -86,8 +86,8 @@ type ConfigMapOptions struct {
 	FileSources []string
 	// LiteralSources to derive the configMap from (optional)
 	LiteralSources []string
-	// EnvFileSource to derive the configMap from (optional)
-	EnvFileSource string
```
Comment (eddiezane) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap.go:90]: This struct and field are exported. This will be a breaking change to any external consumers. Not sure if there are any 🤔.

https://cs.k8s.io/?q=EnvFileSource&i=nope&files=&excludeFiles=&repos=

Comment (eddiezane) [reply staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap.go:90]: I guess with our refactoring of the commands anyways it doesn't matter.

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go [hunk 5]
```diff
@@ -33,11 +33,12 @@ func TestCreateConfigMap(t *testing.T) {
...
 		fromFile      []string
-		fromEnvFile   string
+		fromEnvFile   []string
 		setup         func(t *testing.T, configMapOptions *ConfigMapOptions) func()
 
-		expected  *corev1.ConfigMap
-		expectErr bool
+		expected     *corev1.ConfigMap
+		expectErr    bool
+		expectErrStr string
```
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go:41]: I'd suggest consolidating into only the string field. We should never expect a blank error, so the boolean is kinda redundant and needs to be synchronized (e.g. right now you could set an error string but forget to set expectErr to true and it wouldn't get checked).

Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go:41]: makes sense

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap.go (hunk 4 above)
Comment (KnVerey) [reply staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap.go:90]: My instinct is that the flags used to build a CLI are particular to that CLI itself, and not intended for library use. That search isn't really showing any consumers in any case; Kustomize's implementation matches but is independent, and driver-registrar is just showing up as vendored, not used. The rest is this code itself and its mirror. So IMO there's no problem with this change, but we can check with other approvers if you're worried about it. :)

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go (hunk 1 above)
Comment (KnVerey) [reply staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go:443]: Thanks for digging into this, and great job with the improvements to this file!

**Review by KnVerey — COMMENTED** [review]
/lgtm

**Comment by eddiezane** [comment]
@lauchokyip sorry I missed this. Please rebase when you get a chance and we can get merged after code thaw.

**Comment by lauchokyip** [comment]
@eddiezane  Rebased. PTAL 😄

**Comment by eddiezane** [comment]
Thanks!

What's the intended behavior for duplicate entries? Second one overrides the first?

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go (hunk 1 above)
Comment (KnVerey) [reply staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go:461]: @lauchokyip your latest rebase seems to have lost the changes from this discussion, i.e. the ones visible in https://github.com/kubernetes/kubernetes/compare/cc37d5df54caaaa50291d37c93042d73ea506778..977ca6feb4fe433e667df051dfa4b4c21e4fce02. Can you please restore them? Note that one of the error strings that shows up in that diff answers @eddiezane question. 😄

**Comment by KnVerey** [comment]
> What's the intended behavior for duplicate entries? Second one overrides the first?

Second one is rejected. Regardless of original source, the entries are added with `addKeyFromLiteralToConfigMap`, which as documented "return[s] an error if the key is not valid or if the key already exists." This behavior didn't need to change from the original single-file implementation, since it was already a possibility there.

**Comment by KnVerey** [comment]
In case my comment is hard to find, I noticed yesterday that the latest force push seems to have lost some of the previous requested changes: https://github.com/kubernetes/kubernetes/pull/101646#discussion_r672556459

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go (hunk 1 above)
Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go:494]: Yes I am restoring them! Thanks for the reminder 😃

**Comment by lauchokyip** [comment]
@KnVerey and @eddiezane , I restored the previous changes, PTAL, thank you

**Comment by finnickkim** [comment]
- [ ]

**Review by KnVerey — COMMENTED** [review]
Just a tiny request to revert an additional change that slipped in, otherwise looks good!

File: staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap.go [hunk 6]
```diff
@@ -68,8 +68,8 @@ var (
 		  # Create a new config map named my-config from the key=value pairs in the file
 		  kubectl create configmap my-config --from-file=path/to/bar
 
-		  # Create a new config map named my-config from an env file
-		  kubectl create configmap my-config --from-env-file=path/to/bar.env`))
+		  # Create a new configmap named my-config from an env file
```
[the remark this answers is not in the document]
Comment (lauchokyip) [reply staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap.go:71]: Will do! Thanks for the fast review!

**Comment by eddiezane** [comment]
Great job team! Awesome work on this one.

/approve
